pkgs/development/tools/build-managers/meson/setup-hook.sh

```diff
@@ -31,7 +31,69 @@ mesonConfigurePhase() {
     runHook postConfigure
 }
 
+mesonBuildPhase() {
+    runHook preBuild
+
+    localFlagsArray=(
+        -j$NIX_BUILD_CORES
+        $mesonBuildFlags "${mesonBuildFlags[@]}"
+    )
+
+    echoCmd 'build flags' "${flagsArray[@]}"
+    TERM=dumb meson compile "${flagsArray[@]}"
+
+    runHook postBuild
+}
+
+mesonCheckPhase() {
+    runHook preCheck
+
+    local checkCores=1
+
+    # Parallel checking is enabled by default.
+    if [ "${enableParallelChecking-1}" ]; then
+        checkCores="$NIX_BUILD_CORES"
+    fi
+
+    localFlagsArray=(
+        -j$checkCores
+        $mesonCheckFlags "${mesonCheckFlags[@]}"
+    )
+
+    echoCmd 'check flags' "${flagsArray[@]}"
+    TERM=dumb meson test "${flagsArray[@]}"
+
+    runHook postCheck
+}
+
+mesonInstallPhase() {
+    runHook preInstall
+
+    # shellcheck disable=SC2086
+    localFlagsArray=($mesonInstallFlags "${mesonInstallFlagsArray[@]}")
+
+    echoCmd 'install flags' "${flagsArray[@]}"
+    TERM=dumb meson install "${flagsArray[@]}"
+
+    runHook postInstall
+}
+
 if [ -z "${dontUseMesonConfigure-}" -a -z "${configurePhase-}" ]; then
     setOutputFlags=
     configurePhase=mesonConfigurePhase
 fi
+
+if [ -z "${dontUseMesonBuild-${dontUseMesonConfigure-}}" -a -z "${buildPhase-}" ]; then
```

Comment:
What if ninja setup hook runs first?

Reply:
That's a very good point.  When I was originally thinking about this, I assumed that the `dontUseNinjaBuild` etc. checks wouldn't happen until their respective phases, so setting them immediately in the meson hook would always run first, regardless of ordering.  But that's not actually how it works, so I'm not sure what to do here.  Any ideas?

Reply:
I suppose one thing we could do would be to, in `mesonConfigurePhase`, do something like

```bash
if [[ $checkPhase = ninjaCheckPhase && -z $dontUseMesonCheck ]]; then
    checkPhase=mesonCheckPhase
fi
```

It doesn't feel as nice, but it would get around the ordering problem, and it wouldn't make it any more difficult to force the ninja version if you needed to because you'd still just `dontUseMesonCheck = true`.

Reply:
Yeah, I do not see many other options. It is still better than [what we do in g-i setup hook](https://github.com/NixOS/nixpkgs/blob/d19f38fda982d507684b65efcb00e4b143597843/pkgs/development/libraries/gobject-introspection/setup-hook.sh#L21-L28).

Reply:
Possibly, we could use an auxiliary variable `usingImplicitNinjaCheckPhase=1` added to ninja setup hook to distinguish it from user explicitly setting `checkPhase=ninjaCheckPhase` but I probably would not bother.